Graph interaction via Chart.JS

AKaudinov opened the pull request:
- Implemented graph Interaction via Chart.js and chartjs plugin zoom
   - Individual graphs can be zoomed and panned
- Created a new component: `LineGraph` to keep things readable and easier to navigate
   - New component contains the Chart.js line graph, its axis options -- including zooming and panning
- Refactored the building out of the x axes series from each individual graph, into the main graphs component, so the x series gets built only when data changes, or the date filter is changed, and is passed down to each graph, instead of each graph re-computing the same exact thing 10 times.
- Added the ability to overlay the graph on the same chart by enabling a secondary Y axes on the same component, instead of rendering A new graph component.
  - This also keeps the zoom and pan of the graph that the overlay -- secondary data -- will be on top of without re-rendering the component but updating its visibility options, in place
- The graph also formats the the labels to be more precise when `day` filter is selected, to avoid showing the same exact day throughout the whole graph as it does today.

- Switched the `Graphs` component to be lazy loaded (dynamically loaded with no SSR) as Chart.js and zoom depend on Browser window API.


- Refactored out no longer necessary Nivo graph options
- Removed Nivo related packages

- Refactored no longer necessary Graph seeker styles that were previously used to make the graph seeker work on the data page
- Refactored no longer necessary conditional functionality and props in the DateRangePicker component as it's no longer used as a date seeker


---
Other:

- Refactored the progress bar animation to use compositor only properties (e.g. Transform and Opacity) for the animation. This avoids repaints
  which has much less load on the CPU, hence increasing
  performance

---

Also closes #153

Commit:
feat: working on a data graph date seeker

- setting up a data graph date seeker
- updating the graph to allow for date seeking functionality
- Updated the date seeker and graphs to reflect the date range selection changes

Commit:
feat: continuing to work on the graph date seeking

- moved dsColumns array outside the Component so it doesn't recreate it
  on each re-render as the values are always the same
- corrected the hooks and seeker value in the Graph Component
  so referential equality checks run properly

Commit:
feat: added ability to seek the overlayed graph together with the original

Commit:
fix: graph under seek not displaying when date changes

- Fixed an issue where the graph under seek would not show data
  on the graph when the date filter changed and the seeker value
  was changed prior to the date filter change

Commit:
feat: continuing to work on the graph seeker

Commit:
refactor: refactored the dispay of the range input

Commit:
chore: added more notes

Commit:
feat: seeker aria props and steps

- Added aria accessibility props to the graph seeker
- Added custom steps to the seeker based on the activeUnit selected
  by the graph date filter

Commit:
chore: removed un-needed comment

Commit:
feat: seeker flex-box and label

- Updated the graph seeker styling and wrapped it in a flex-box
- Added a label for the seeker
- Moved the seeking functionality under the graph where it amkes more sense

Commit:
feat: updating seeker steps

Commit:
feat: removed unnecessary comment and moved conditional rendering below hooks

Commit:
refactor: updated the max value of tide constant

Commit:
refactor: updated label from for to htmlFor

Commit:
merging latest next-13 branch and fixing conflicts

Commit:
feat: overwriting rc-slider styles within module

- Working on overwriting rc-slider styles
  within the dataRangePicker module css file
  so we don't overrwrite them globally but only withn the module
   - this provides the ability for the plan to use the rc-slider
     on the data page when zooming graph data.

Commit:
merge branch 'next-13' into graph-seeker

Commit:
feat: Stop gap: DateRangePicker global classes restyling and overrides

- Removed the global rc-slider css overrides and moved them over
  to the DateRangePicker css module
   - Updated the DateRangePicker css module to properly override
     the rc-slider styles

- Updated the date graph seeker On the Data pate to use
  the DateRangePicker component for re-usability without index styles

- Updated the DateRangePicker with an optional prop that determines
  whether it is being rendered by the index page, if so, it applies
  all the necessary styles to properly display on the index page.

Commit:
refactor: updated the padding of the graphs top bar

Commit:
feat: data page rc-slider functionality

- Set up specific rc-slider style overrides for the data page
- Updated the DataRangePicker to dynamically apply the styles
  whether it is on the data or index page
- Updated the Rc-slider seeker container to have a class name
  instead of inline styles
- Updated the data page to use the customized rc-slider and re-use
  the DataRangePicker component with the custom values from each graph

Commit:
refactor: removed unnecessary code and props

- Removed unnecessary code from graph component that was
  setting up steps for the range input which is no longer used
- Removed unnecessary activeUnit prop from Graph

Commit:
doc: added more notes and comments

Commit:
doc: updated graph seeker height and removed unnecessary comments

Commit:
refactor: updated the height of the rail track on data page

Commit:
feat: graph seeking maximum available date to avoid holes in graph

- Updated the Graph x scale to set a maximum date value -- left hand side --
  to be the latest available date from the cut data, this avoids holes in the data
  when the latest available date from data has a gap between the current seek date
  and presents a hole in the graph

Commit:
refactor: removed legends from underneath certain graphs

Commit:
doc: added contextual comments

Commit:
merge branch 'next-13' into graph-seeker

Commit:
added npmrc

Commit:
Merge branch 'stable' into graph-seeker

Commit:
refactor: removed unnecessary code and added contextual comments

Commit:
feat: working on chart js and zoom + pan

- Experimenting and trying to implement chart js
  with zooming and panning functionality

Commit:
refactor: changed the graph labeling and formatting

- removed the time scale x series as the minimum
  zoom level wasn't being respected
- Switched to a function that builds x and y series separately
- Updated the graph labeling so each unix timestamp on an x axis
  is labeled as a month, day, year
- Updated the tooltip title to display date and time
- Updated the tooltip label to display a value using a fixed-point
  notation, along with the graph unit

Commit:
feat: continued work on the charts, panning and zooming

- Refactored the LineGraph component to include useEffects and memoization
  in order to not recalculate the same data and graph options upon any non
  related changes
- Updated the LineGraph component with its own styles
- Added onMouseDown/Up events to change the dragging cursor
- Refactored the styling of the LineGraph component
- Refactored the graph styling of the charts to be more readable

Commit:
Stop gap: continuing to work on the graphs and their props

Commit:
Stop gap: continuing to work on chartjs and graphs

- Implemented the building of x series on the graphs component
  so it builds only once instead of for each graph
  - Refactored the graphs to work with the new x series labels
- Refactored the setup of the overlay y series data so it forms
  as the component is about to render, not after it does.
  This prevents the jumping of the overlay data when switching
  date filters, as it "lags" behind
- Refactored and removed animations from chart.js to further
  improve performance on the down sampled data

Commit:
Continuing research into chartjs

Commit:
feat: refactored the progress bar animation

- Switched to transform animation that avoids repaints
  which has much less load on the CPU, hence increasing
  performance

Commit:
removed unnecessary commented code

Commit:
switched to day as the default date unit

Commit:
feat: implementing chartjs for graph interaction

- Added functionality to change graph labels -- padding and display format --  based on the
  selected date unit from the graph date filter

Commit:
Refactor: Removing code and updating graph tooltip data

- Removed unnecessary code from the Graph and Graphs components
- Updated the linechart configuration to use proper data units
  when displaying a tooltip with an overlayed dataset

Commit:
Stop gap: refactoring functionality and naming

Commit:
refactor: removed unnecessary code and refactoring

- Refactored the date time units into an "enum"
- Removed unnecessary code from DateRangePicker that was related to the data page
- Removed unnecessary data page styles from the DateRangePicker

Commit:
fix: fixed an issue where each about Section didn't have a unique key

Commit:
refactor: Improving performance and usability of the graphs

- Updated the LineGraph component to only recompute the
  chart options when the date filter changes, instead of each time
  causing the chart to reset to default settings even when
  and overlay graph is placed
   - New functionality allows for placing an overlay graph
     and keeping existing chart options by updating the chart
     in place without rerendering
   - Updated the Graph component to pass in a new `hasOverlay` prop
     to the LineGraph component, to keep the necessary referencial working
     properly when an overlay graph is added, without depending on the
     overlayGraph object and doing shallow comparisons.
- Removed unnecessary code from the Graph component
- Removed ChartJs registration from each LineGraph component
- Updated the Graphs component to rebuild the x axes series
  not only when the date filter active unit changes, but also
  if new data is present

Commit:
doc: adding contextual notes

Commit:
Removed unnecessary packages:
- Nivo/core
- Nivo/line
- Chartjs-dayjs-adapter

Commit:
Added aria labels to graph buttons

Commit:
Switched to default date unit of week

Commit:
Added more contextual comments

Commit:
doc: more comments

jedahan commented on components/DataRangePicker/index.js:
```diff
@@ -101,10 +102,12 @@ class DataRangePicker extends React.Component {
             })}
           />
         </div>
+        {(
```
some debugging leftover?

AKaudinov replied:
Ah, thanks for catching that! forgot to remove those parentheses after removing the conditional.

Commit:
removed no longer necessary jsx parser curly braces

jedahan commented on components/Graph/index.js:
```diff
@@ -135,25 +74,17 @@ const Graph = ({
+          activeUnit={activeUnit}
+          xSeries={xSeries}
+          data={graph.data}
+          domain={[graph.domain[0], graph.domain[1]]}
```
will be removed

jedahan commented on components/DataRangePicker/index.js:
```diff
@@ -85,8 +85,9 @@ class DataRangePicker extends React.Component {
     return (
       <div className={styles.container}>
-        <div className={styles.sliderContainer}>
+        <div className={`${styles.sliderContainer} ${styles.customSliderContainer}`}>
```
so we can `[styles.sliderContainer, styles.customSliderContainer]`

Commit:
refactor: removed no longer used domain prop from graphs

Commit:
refactor: removed deprecated formxyseries method

Commit:
refactor: updated daterangepicker with cx classname function

- using a cx function to add multiple classes instead of template literals

Commit:
refactor: refactored activeUnit to activeDateFilter

- Updated all activeUnit references to activeDateFilter
- Updated setActiveUnit to setActiveDateFilter
